### PR TITLE
Selenium 4

### PR DESCRIPTION
**Pull request by tcnh**
New Selenium 4 PR. (continues on base of https://github.com/fhoeben/hsac-fitnesse-fixtures/pull/355)

Selenium 4.1.2 version
Relative locators
Shadow DOM access by using existing searchcontexts in Chromium 96+
DevTools Browsertest as a sample
ChromiumOptions for Chrome- and Edgedriver, with examples wiki page

Includes acceptance tests for shadowdom and relative locators

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -43,8 +43,8 @@
 		<hsac-fitnesse-plugin.version>1.31.0</hsac-fitnesse-plugin.version>
 		<fitnesse.version>20211030</fitnesse.version>
 		<praegus-toolchain-fitnesse-plugin.version>2.0.14</praegus-toolchain-fitnesse-plugin.version>
-		<selenium.version>3.141.59</selenium.version>
-		<slf4j.version>1.7.32</slf4j.version>
+		<selenium.version>4.1.2</selenium.version>
+		<slf4j.version>1.7.25</slf4j.version>
```
Comment (fhoeben) [inline pom.xml:47]: Any particular reason to downgrade slf4j?

File: wiki/FitNesseRoot/HsacExamples/SlimTests/BrowserTests/Selenium4Examples.wiki [hunk 2]
```diff
@@ -0,0 +1,3 @@
+!1 Test suite X
```
Comment (fhoeben) [inline wiki/FitNesseRoot/HsacExamples/SlimTests/BrowserTests/Selenium4Examples.wiki:1]: I suppose this file is no longer used?

**Review by fhoeben — COMMENTED** [review]
Haven't gone in full details yet, but looks pretty good.

Some global thoughts:
- should we add some documentation (references) to the top-level examples.BrowserTest.wiki for the new commands/features?
- the relative locators add a lot of new methods to the (already too big) BrowserTest class. And the require technical selectors (which you very nicely explicitly documented). Can we do something about that? I'm thinking whether it would be bad to have them in a separate subclass (I like that they are available, but think I would prefer tests that do not (need to) use them). Do you have any thoughts?

File: pom.xml (hunk 1 above)
Comment (tcnh) [reply pom.xml:47]: was meant to be .35 but I think i typoed. Will update

**Comment by tcnh** [comment]
> Haven't gone in full details yet, but looks pretty good.
> 
> Some global thoughts:
> 
> * should we add some documentation (references) to the top-level examples.BrowserTest.wiki for the new commands/features?

I added them to the commands table (value of <> above/below/leftof/right of/near was missing, is added now)
I feel giving them much more acknowledgement would be to much credit for methods that should relatively not be used very often?

> * the relative locators add a lot of new methods to the (already too big) BrowserTest class. And the require technical selectors (which you very nicely explicitly documented). Can we do something about that? I'm thinking whether it would be bad to have them in a separate subclass (I like that they are available, but think I would prefer tests that do not (need to) use them). Do you have any thoughts?

I agree (see above) that these are a bit of niche. On the other hand this is the most prominent addition to Selenium 4 for most end-users and I also believe there are a whole lot of methods in BrowserTest that are only used in specific use cases, so these are  not that different. Splitting browsertest into multiple layers (of complexity/technology orientation/etc) might be a good idea, but as I mentioned, I believe we could move out a whole bunch of other methods and settings, too.

**Pull request merged**